bin/driver.ml

```diff
@@ -36,140 +22,171 @@ let typecheck_only
+      = fun Loader.({ context; program_ }) ->
+      Frontend.interactive context program_
+  end
+
+  module Compile = struct
+    module IR = struct
+      let run : Sugartypes.program Frontend.result -> Sugartoir.result
+        = fun Frontend.({ context; datatype; program }) ->
+        Sugartoir.program context datatype program
+    end
+  end
+
+  module Transform = struct
+    let run : Sugartoir.result -> Backend.result
+      = fun Sugartoir.({ context; datatype; globals; program }) ->
+      let program' =
+        Ir.with_bindings globals program (* TODO(dhil): This looks silly, and I think it may be wrong... *)
+      in
+      Backend.program context datatype program'
+  end
 
-      let globals = Backend.transform_prelude tenv globals in
+  module Evaluate = struct
+    module Eval = Evalir.Eval(Webserver)
 
-      let valenv = Eval.run_defs Value.Env.empty globals in
-      let envs =
-        (valenv,
-        Env.String.extend Lib.nenv nenv,
-        Types.extend_typing_environment Lib.typing_env tyenv)
+    let run : Backend.result -> (Context.t * Types.datatype * Value.t)
+      = fun Backend.({ context; program; datatype }) ->
+      let valenv = Context.value_environment context in
+      let (valenv', v) =
+        lazy (Eval.run_program valenv program) |>measure_as<| "run_program"
       in
-        globals, envs
-   in
-   die_on_exception load_prelude_inner ()
+      (Context.({ context with value_environment = valenv' }), datatype, v)
+  end
+
+  module Interactive = struct
+    type result =
+      { context: Context.t;
+        sentence: [ `Definitions of Ir.binding list
+                  | `Expression of Types.datatype * Value.t
+                  | `Directive of string * string list ] }
+
+    let readline : string -> Context.t -> result
+      = fun ps1 context ->
+      let Frontend.({ context; program = sentence; datatype })
+        = Parse.interactive ps1 context
+          |> Desugar.interactive
+      in
+      let context, sentence =
+        match sentence with
+        | Sugartypes.Definitions defs ->
+           let program' = (defs, None) in
+           let Backend.({ program; _ }) as result =
+             Compile.IR.run Frontend.({ context; datatype; program = program' })
+             |> Transform.run
+           in
+           let (context', _, _) = Evaluate.run result in
+           context', `Definitions (fst program)
+        | Sugartypes.Expression expr ->
+           let program' = ([], Some expr) in
+           let result =
+             Compile.IR.run Frontend.({ context; datatype; program = program' })
+             |> Transform.run
+           in
+           let (context', datatype, value) = Evaluate.run result in
+           context', `Expression (datatype, value)
+        | Sugartypes.Directive d ->
+           context, `Directive d
+      in
+      { context; sentence }
+  end
+
+  let dump_lib : out_channel -> unit
```

Comment:
It seems like `dump_lib`, `initialise`, `whole_program`, and `evaluate_string` shouldn't be part of `Phases`, but go into `Driver` directly